Space Ninjas auto-toggle internals after spawning

Errant-4 opened the pull request:
## About the PR
Space ninja masks automatically toggle internals on at roundstart (technically, when they are first equipped. Thus, space ninja masks spawned with F5 will do this the first time they are worn)

## Technical details
When an entitiy with `BreathToolComponent.AutomaticActivation = true` is equipped, it ~~schedules the werarer's internals to be toggled on the next tick. (The delay is to ensure that the ninja's gear finished spawning and thus there is an air tank to work with)~~ immediately toggles the wearer's internals.
Currently, this field is then set to false, so this only happens the first time when such a mask is equipped.

To ensure that this can actually work at spawn (where the mask slot item is normally spawned _before_ the suitstorage item), a new passable parameter in InventorySystem.TryGetSlots reorders the output so the Mask slot will be listed last. This parameter is only used in the spawning system, so other systems requesting the inventory slots and relying in them to be in a certain order (such as item smartequip) are not impacted

## Media
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl: Errant
- tweak: Space Ninjas now automatically turn their internals on when they spawn.

Jezithyr commented on Content.Server/Body/Systems/InternalsSystem.cs:
```diff
@@ -38,6 +40,38 @@ public override void Initialize()
+                return;
+
+            if (TryComp<InternalsComponent>(uid, out var comp))
+                ToggleInternals(delayComp.Entity , delayComp.Entity , false, comp);
```
You should probably check that the internals are not already active before toggling them on. That shouldn't be the case but it may prevent future bugs/unexpected behavior.

Errant-4 replied:
I figured such a check wouldn't be needed since internals being connected means it has just been reset, but you are right. It's better to make things more future-safe

Jezithyr commented:
Pretty good quality of life change, other than that one nitpick it looks good to me!

metalgearsloth commented on Content.Server/Atmos/Components/BreathToolComponent.cs:
```diff
@@ -16,5 +16,12 @@ public sealed partial class BreathToolComponent : Component
         public SlotFlags AllowedSlots = SlotFlags.MASK | SlotFlags.HEAD;
         public bool IsFunctional;
         public EntityUid? ConnectedInternalsEntity;
+
+        /// <summary>
+        /// Tool will automatically turn on internals when it is next equipped
+        /// This feature will turn itself off after the first time it's used
+        /// </summary>
+        [DataField("automaticActivation")]
+        public bool AutomaticActivation = false;
     }
```
Datafield strings are obsolete.

```suggestion

        /// <summary>
        /// Tool will automatically turn on internals when it is next equipped
        /// This feature will turn itself off after the first time it's used
        /// </summary>
        [DataField]
        public bool AutomaticActivation = false;
    }
```

metalgearsloth commented on Content.Server/Body/Systems/InternalsSystem.cs:
```diff
@@ -38,6 +40,38 @@ public override void Initialize()
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<InternalsDelayedActivationComponent>();
+        while (query.MoveNext(out var uid, out var delayComp))
+        {
+            if (_timing.CurTime < delayComp.Time)
+                return;
+
```
```suggestion
        var query = EntityQueryEnumerator<InternalsDelayedActivationComponent>();
        while (query.MoveNext(out var uid, out var delayComp))
        {
            if (_timing.CurTime < delayComp.Time)
                continue;

```

return will break for > 1 mask.

metalgearsloth commented on Content.Server/Body/Systems/InternalsSystem.cs:
```diff
@@ -261,3 +298,28 @@ private short GetSeverity(InternalsComponent component)
         return null;
     }
 }
+
+/// <summary>
+/// Raised on a breathing mask entity after it was equipped on a mob and successfully connected to the mob's InternalsComponent
+/// </summary>
+public sealed class ConnectedBreathToolEvent : EntityEventArgs
+{
+    /// <summary>
+    /// The mask that was equipped
+    /// </summary>
+    public readonly EntityUid Mask;
+
+    /// <summary>
+    /// The mob that just equipped the mask
+    /// </summary>
+    public readonly EntityUid User;
+
+    public readonly InternalsComponent IntComp;
+
+    public ConnectedBreathToolEvent (EntityUid mask, EntityUid user,  InternalsComponent intComp)
+    {
+        Mask = mask;
+        User = user;
+        IntComp = intComp;
+    }
+}
```
All new events need to be record structs.

metalgearsloth commented on Content.Server/Body/Systems/InternalsSystem.cs:
```diff
@@ -38,6 +40,38 @@ public override void Initialize()
+            RemCompDeferred<InternalsDelayedActivationComponent>(uid);
+        }
+    }
+
+    private void OnBreathMaskConnected(EntityUid uid, BreathToolComponent comp, ConnectedBreathToolEvent args)
+    {
+        if (!comp.AutomaticActivation)
+            return;
+
+        // delaying until next tick, because at spawn we can't be sure the gas tank will spawn before the mask
+        EnsureComp<InternalsDelayedActivationComponent>(args.User, out var delayComp);
+        delayComp.Time= _timing.CurTime + TimeSpan.FromTicks(1);
+        delayComp.Entity = args.User;
+
+        // this is currently only used for roundstart activation, so we turn the feature off afterwards
+        comp.AutomaticActivation = false;
     }
```
Isn't this going to be really janky in gameplay? It's also not predicted.

Errant-4 replied:
I'm not sure how/which part of this is janky gameplay-wise? Can you please be more specific

Errant-4 replied:
Well the one part I see as janky, currently, is that there is a doafter for the internal turning on. But I didn't want to duplicate parts of the toggle code to be able to do it without the doafter. I'm thinking I'll put an "Instant" bool there that allows calling the same classes and just skips over the doafter

Errant-4 replied:
The doafter for toggling internals was removed on master so that's no longer an issue for auto-activation either

Errant-4 replied:
It should also be predicted now

metalgearsloth commented on Content.Server/Atmos/Components/InternalsDelayedActivationComponent.cs:
```diff
@@ -0,0 +1,20 @@
+﻿namespace Content.Server.Atmos.Components;
+
+/// <summary>
+/// This entity will try to toggle their internals at the specified time
+/// This component is automatically created and deleted by InternalsSystem
+/// </summary>
+[RegisterComponent]
+public sealed partial class InternalsDelayedActivationComponent : Component
+{
+    /// <summary>
+    /// The entity that will toggle internals
+    /// </summary>
+    public EntityUid Entity = new EntityUid();
+
+    /// <summary>
+    /// The server time when the internals will be toggled
+    /// </summary>
+    public TimeSpan Time = new TimeSpan();
+}
+
```
You shouldn't be constructing the structs here, also no prediction or datafields which are required.

Errant-4 replied:
What does prediction do on a server component? I thought that's for the client

deltanedas replied:
it should be in shared and predicted

Errant-4 replied:
Got it, thanks

metalgearsloth reviewed (changes requested):
Thinking about it this really just seems like hacks around implementing startinggear ordering if it's just going to activate it the next 'tick'. There's no design reason for this to even exist besides "I'm not sure on entity spawning" which we can guarantee for anything that currently cares about this.

Errant-4 commented:
> Thinking about it this really just seems like hacks around implementing startinggear ordering if it's just going to activate it the next 'tick'. There's no design reason for this to even exist besides "I'm not sure on entity spawning" which we can guarantee for anything that currently cares about this.

I removed the whole delay thing and introduced a new bool to the class used to get an entity's inventory slots. When this flag is set on the request, "mask" is moved to the end of the list of slots (if present)
This works for the current scenario, and is technically "done".

But is that enough? This is a very specific transformation with no future potential.
Should I replace it with a more complex functionality that can take some sort of rule list/template for future uses, and pass such a rule from the gear spawning code, rather than the current book?

metalgearsloth commented:
I have big brained a solution but will fix this when I get time.
